packages/docgen/lib/get-type-annotation.js

```diff
@@ -0,0 +1,424 @@
+	}
+	return '';
+}
+
+/**
+ * @param {babelTypes.TSTypeLiteral} typeAnnotation
+ */
+function getTypeLiteralIndexSignatureTypeAnnotations( typeAnnotation ) {
+	const indexers = typeAnnotation.members
+		.map( ( m ) => babelTypes.isTSIndexSignature( m ) )
+		.map( ( indexer ) => {
+			const parameter = indexer.parameters[ 0 ];
+			return `[ ${ parameter.name }: ${ getTypeAnnotation(
+				parameter.typeAnnotation
+			) } ]: ${ getTypeAnnotation( indexer.typeAnnotation ) }`;
+		} );
+
+	if ( indexers.length ) {
+		return `${ indexers.join( '; ' ) } `;
+	}
+	return '';
+}
+
+/**
+ * @param {babelTypes.TSTypeLiteral} typeAnnotation
+ */
+function getTypeLiteralPropertyTypeAnnotations( typeAnnotation ) {
+	const properties = typeAnnotation.members
+		.filter( ( m ) => babelTypes.isTSPropertySignature( m ) )
+		.map( ( prop ) => {
+			return `${ prop.key.name }: ${ getTypeAnnotation(
+				prop.typeAnnotation
+			) }`;
+		} );
+
+	if ( properties.length ) {
+		return `${ properties.join( '; ' ) } `;
+	}
+	return '';
+}
+
+/**
+ * @param {babelTypes.TSTypeLiteral} typeAnnotation
+ */
+function getTypeLiteralTypeAnnotation( typeAnnotation ) {
+	const callProperties = getTypeLiteralCallSignatureDeclarationTypeAnnotations(
+		typeAnnotation
+	);
+	const indexers = getTypeLiteralIndexSignatureTypeAnnotations(
+		typeAnnotation
+	);
+	const properties = getTypeLiteralPropertyTypeAnnotations( typeAnnotation );
+
+	return `{ ${ callProperties }${ indexers }${ properties }}`;
+}
+
+/**
+ * @param {babelTypes.TSUnionType} typeAnnotation
+ */
+function getUnionTypeAnnotation( typeAnnotation ) {
+	return typeAnnotation.types.map( getTypeAnnotation ).join( ' | ' );
+}
+
+/**
+ * @param {babelTypes.TSIntersectionType} typeAnnotation
+ */
+function getIntersectionTypeAnnotation( typeAnnotation ) {
+	return typeAnnotation.types.map( getTypeAnnotation ).join( ' & ' );
+}
+
+/**
+ * @param {babelTypes.TSArrayType} typeAnnotation
+ * @return {string} The type annotation
+ */
+function getArrayTypeAnnotation( typeAnnotation ) {
+	return `${ getTypeAnnotation( typeAnnotation ) }[]`;
+}
+
+/**
+ * @param {babelTypes.TSTupleType} typeAnnotation
+ */
+function getTupleTypeAnnotation( typeAnnotation ) {
+	const types = typeAnnotation.elementTypes
+		.map( getTypeAnnotation )
+		.join( ', ' );
+	if ( types.length ) {
+		return `[ ${ types.join( ', ' ) } ]`;
+	}
+	return '[]';
+}
+
+/**
+ * @param {babelTypes.TSQualifiedName} qualifiedName
+ */
+function unifyQualifiedName( qualifiedName ) {
+	if ( ! qualifiedName.right ) {
+		if ( ! qualifiedName.left ) {
+			return qualifiedName.name;
+		}
+		return qualifiedName.left.name;
+	}
+	return `${ unifyQualifiedName( qualifiedName.left ) }.${
+		qualifiedName.right.name
+	}`;
+}
+
+/**
+ * @param {babelTypes.TSImportType} typeAnnotation
+ */
+function getImportTypeAnnotation( typeAnnotation ) {
+	// Should this just return the unqualified name (i.e., typeAnnotation.name || typeAnnotation.right.name)?
```

Comment:
This is still an outstanding question to me... I think we should just use the unqualified name without the import, for stylistic reasons, but maybe using the import would be best. In any case, import types are rare in actual TypeScript so it wouldn't happen often. The key is to use descriptive and contextually appropriate type names anyway.